### ISSUE TITLE
text slicer out of memory error

### ISSUE THREAD

**Issue by rcrath** [Jermolene/TiddlyWiki5#3570]
Node.js TW5.1.17 on manjaro, 8 GB memory up to date with text slicer plugin.  TW is as follows:
* 2,183 items, totalling 298.6 MB, 5.8 MB in text-based tiddlers
* 42 items are image and audio, totalling 292.2 MB
* 1,023 base tiddlers
* Sliced about 5-10 large tiddlers full of book notes
* that made 1,116 sliced items, totaling 583.3 kB

THis does not look sustainable as it will produce thousands of inodes...maybe sliced texts should have their own subdirectories.

BTW, found a neat solution to keeping them organized but out of the way...naming them with the name of the original tiddler prefixed with an Omega (Ω) sorts them all after z, keeping them out of the way in searches, which will find stuff in the original text first.  
 
to reproduce
* slice large tiddlers on 16 gig machine
* load TW on 8 gig machine
* get following error on command line, with no load of course on the browser:

```
rich[~]$ tiddlywiki /home/rich/Dropbox/FiddlyWiki --server
Serving on 127.0.0.1:8080
(press ctrl-C to exit)
 syncer-server-filesystem: Dispatching 'save' task: $:/StoryList 
 filesystem: Saved file /home/rich/Dropbox/FiddlyWiki/tiddlers/$__StoryList.tid 

<--- Last few GCs --->

[2700:0x5594b1669210] 24418040 ms: Mark-sweep 901.1 (945.3) -> 901.0 (914.3) MB, 79.4 / 0.0 ms  (average mu = 1.000, current mu = 0.000) last resort GC in old space requested
[2700:0x5594b1669210] 24418128 ms: Mark-sweep 901.0 (914.3) -> 901.0 (914.3) MB, 87.3 / 0.0 ms  (average mu = 1.000, current mu = 0.000) last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x2841df64fb5d]
Security context: 0x3ed38631d949 <JSObject>
    1: DoJoin(aka DoJoin) [0x12e84a924979] [native array.js:~89] [pc=0x2841df6de0a9](this=0x3bc4fb3825b1 <undefined>,0x022e30cd3069 <JSArray[36]>,36,0x3bc4fb382741 <true>,0x3bc4fb382851 <String[0]: >,0x3bc4fb382801 <false>,0x3bc4fb3825b1 <undefined>,0x3bc4fb3825b1 <undefined>)
    2: Join(aka Join) [0x12e84a9249c9] [native array.js:1] [bytecode=0x18f1454e309...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: 0x5594af5ea361 node::Abort() [node]
 2: 0x5594af5ea3af  [node]
 3: 0x5594af79c842 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0x5594af79ca9b v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0x5594afb323a3  [node]
 6: 0x5594afb44d15 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
 7: 0x5594afb1314a v8::internal::Factory::AllocateRawWithImmortalMap(int, v8::internal::PretenureFlag, v8::internal::Map*, v8::internal::AllocationAlignment) [node]
 8: 0x5594afb1a551 v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [node]
 9: 0x5594afdce410 v8::internal::Runtime_StringBuilderConcat(int, v8::internal::Object**, v8::internal::Isolate*) [node]
10: 0x2841df64fb5d 
Aborted (core dumped)
rich[~]$ 
```

**Comment by Jermolene** [comment]
Hi @rcrath I'm delighted you're experimenting with text-slicer, it's great to get feedback.

It sounds possible that a big part of the problem is the 292MB of images and audio. In the upcoming 5.1.18 release you could shift those to become file attachments, rather than tiddlers:

https://tiddlywiki.com/prerelease/#Using%20the%20integrated%20static%20file%20server

**Comment by Jermolene** [comment]
Hi @rcrath 

The instructions for installing the prerelease under Node.js are here: 
https://tiddlywiki.com/prerelease/#Installing%20TiddlyWiki%20Prerelease%20on%20Node.js

Please do go ahead and create a new thread for any new text-slicer issues.

**Comment by rcrath** [comment]
1. moved audio and large image files to ./tiddlers/files
2. installed and launched TW5 from git successfully.
3. Same error.  

I am back home on my 16 gig machine instead of the 8 Gb laptop.  It was working fine when I left but now gives the same error.  It gave the error before I switched to git, too (i.e. TW5 5.1.17)

Here is the cmd line running up-to-date manjaro linux, same as laptop.

Thanks for looking at this.
I'd be happy to send the tiddlers folder file list if that would shed any insight
```
rich[TiddlyWiki5]$ tiddlywiki /home/rich/Dropbox/FiddlyWiki --listen
Serving on http://127.0.0.1:8080
(press ctrl-C to exit)
 syncer-server-filesystem: Dispatching 'save' task: $:/StoryList 
 filesystem: Saved file /home/rich/Dropbox/FiddlyWiki/tiddlers/$__StoryList.tid 

<--- Last few GCs --->

[12965:0x564aed2a9210]     9508 ms: Mark-sweep 906.3 (951.9) -> 906.2 (919.9) MB, 48.8 / 0.0 ms  (average mu = 0.931, current mu = 0.000) last resort GC in old space requested
[12965:0x564aed2a9210]     9571 ms: Mark-sweep 906.2 (919.9) -> 906.2 (919.9) MB, 63.5 / 0.0 ms  (average mu = 0.843, current mu = 0.000) last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x3b3aa69cfb5d]
Security context: 0x25524ef1d949 <JSObject>
    1: DoJoin(aka DoJoin) [0x33254500dc59] [native array.js:~89] [pc=0x3b3aa6c97e29](this=0x22f0aa6025b1 <undefined>,0x0e2433383881 <JSArray[38]>,38,0x22f0aa602741 <true>,0x22f0aa602851 <String[0]: >,0x22f0aa602801 <false>,0x22f0aa6025b1 <undefined>,0x22f0aa6025b1 <undefined>)
    2: Join(aka Join) [0x33254500dca9] [native array.js:1] [bytecode=0x7db6dbad011...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: 0x564aebb0a361 node::Abort() [node]
 2: 0x564aebb0a3af  [node]
 3: 0x564aebcbc842 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0x564aebcbca9b v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0x564aec0523a3  [node]
 6: 0x564aec064d15 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
 7: 0x564aec03314a v8::internal::Factory::AllocateRawWithImmortalMap(int, v8::internal::PretenureFlag, v8::internal::Map*, v8::internal::AllocationAlignment) [node]
 8: 0x564aec03a551 v8::internal::Factory::NewRawTwoByteString(int, v8::internal::PretenureFlag) [node]
 9: 0x564aec2ee410 v8::internal::Runtime_StringBuilderConcat(int, v8::internal::Object**, v8::internal::Isolate*) [node]
10: 0x3b3aa69cfb5d 
Aborted (core dumped)
rich[TiddlyWiki5]$ 

```

**Comment by BurningTreeC** [comment]
Hi @rcath, I've been running the current prerelease for the last week or so and kept it up-to-date constantly, without issues. Could you provide a command line example how you run it so that more people can test with a similar setting?

**Comment by rcrath** [comment]
Hi @BurningTreeC, the cmd line is the first line of the code excerpt above.
`tiddlywiki /home/rich/Dropbox/FiddlyWiki --listen`

**Comment by rcrath** [comment]
@Jermolene moving the files folder brought the wiki back up.  It looks like I need to manually re-link the moved files.  Having trouble with that.
* This works: `[[http://localhost:8080/files/Georgia Sea Island Singers - Knee-bone - cultural equity T923R14.mp3]]`
* But not this: `[[./files/Georgia Sea Island Singers - Knee-bone - cultural equity T923R14.mp3]]`
* or this: `{{./files/Georgia Sea Island Singers - Knee-bone - cultural equity T923R14.mp3}}`

and oddly, the image files continue to work after they have been moved with no `./files` prefix.  Also, should *.meta files follow the image files to the new files folder or stay put in the tiddlers folder?

But I'm glad to get the wiki back up and operational so I can write tomorrow! thnx!

**Comment by rcrath** [comment]
also, lazy loading does not seem to work with the new --listen command, but still works with --server

@BurningTreeC the command which is working now is `tiddlywiki /home/rich/Dropbox/FiddlyWiki --server $:/core/save/lazy-all'

**Comment by BurningTreeC** [comment]
Hi @rcrath , for lazy-loading have you tried the `root-tiddler` parameter `--root-tiddler="$:/core/save/lazy-all"` ?

As far as I know, embedding external images works using the `[img[path]]` syntax

for links to files, does `[ext[./files/...]]` work?

to embed external files, best practice is to use the `_canonical_uri` field

- create the field `_canonical_uri`
- insert `/files/Georgia Sea Island Singers - Knee-bone - cultural equity T923R14.mp3` as its value
- set the tiddler type to `audio/mp3`

**Comment by rcrath** [comment]
@BurningTreeC using `--root-tiddler="$:/core/save/lazy-all"` fails to start under 5.1.18.  The way I had it, using just the path with no `root=` option, and `--server` rather than the newer `--listen` is the only way I have tried that loads the wiki.  is there a way to check if lazy loading is in fact on?  the `[ext[PATH]]` method worked.  Thank you for reading the manual to me! I looked at that page but failed to register the ext thing. :)

What am I embedding an external URL in? a pre-existing tiddler? or a new one dedicated to the file?
